### PR TITLE
Review HTTP/2 GOAWAY handling

### LINKED ISSUE #5310: Jetty Http2 client discards the response fames when there is GOAWAY and sends RST_STREAM
**Jetty version**  9.4.27.v20200227

**Java version**  OpenJdk 14

**OS type/version** Linux master 4.1.12-112.16.4.el7uek.x86_64 /Deployed in Kubernetes cluster

**Description** 
We are using Jetty Http2 client in our application. We have a scenario where the server will terminate the connection by sending GOAWAY frame with no error along with the response header and data frames. But Jetty client cancels the request by sending RST_STREAM  and terminates the connection by sending GOAWAY. In the application side client also throws "java.nio.channels.AsynchronousCloseException" without committing response.

{"thread":"@d8d9199-4177","level":"ERROR","loggerName":"Exception","message":"Internal Server Error","thrown":{"commonElementCount":0,"name":"java.nio.channels.AsynchronousCloseException","extendedStackTrace":[{"class":"org.eclipse.jetty.http2.client.http.HttpConnectionOverHTTP2","method":"close","file":"HttpConnectionOverHTTP2.java","line":144,"exact":false,"location":"http2-http-client-transport-9.4.27.v20200227.jar!/","version":"9.4.27.v20200227"}],"suppressed":[{"commonElementCount":0,"localizedMessage":"\nError has been observed at the following site(s):\n\t|_ checkpoint ⇢ Request to POST
Capture
Capture

### PR DESCRIPTION
**Pull request by sbordet**

### CODE REVIEW

File: jetty-http2/http2-client/src/main/java/org/eclipse/jetty/http2/client/HTTP2ClientConnectionFactory.java [hunk 1]
```diff
@@ -92,20 +92,6 @@ private HTTP2ClientConnection(HTTP2Client client, ByteBufferPool byteBufferPool,
...
-            return session.getStreamsOpened();
-        }
-
-        @Override
-        public long getMessagesOut()
-        {
-            HTTP2ClientSession session = (HTTP2ClientSession)getSession();
-            return session.getStreamsClosed();
-        }
-
```
Comment (gregw) [inline jetty-http2/http2-client/src/main/java/org/eclipse/jetty/http2/client/HTTP2ClientConnectionFactory.java:108]: why were these removed?

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Flusher.java [hunk 2]
```diff
@@ -365,7 +365,7 @@ protected void onCompleteFailure(Throwable x)
         // If the failure came from within the
         // flusher, we need to close the connection.
         if (closed == null)
-            session.abort(x);
+            session.onWriteFailure(x);
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Flusher.java:368]: comment is not correct as you are not directly closing.
How is onWriteFailure different to abort?

**Review by gregw — COMMENTED** [review]
might get you to walk me through the new tests tomorrow...

**Review by gregw — APPROVED** [review]
LGTM

**Comment by gregw** [comment]
But don't merge until 9.4.35 is promoted

**Review by lorban — APPROVED** [review]
LGTM

**Pull request merged**